java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -23,219 +23,258 @@
-        if (learnedConstraints.anyMatch(terminateTraversal)) {
-          continue;
-        }
-      }
-      for (ExplodedGraph.Node parent : currentNode.parents()) {
-        workList.push(new NodeSymbol(parent, newTrackSymbol));
+      ExecutionPath path = workList.pop();
+      if (path.finished) {
+        flows.add(path.flow);
+      } else {
+        path.lastEdge.parent.edges().stream()
+          .filter(path::notVisited)
+          .map(path::addEdge)
+          .forEach(workList::push);
       }
     }
+    return flows;
   }
 
-  private Stream<Constraint> addFlowFromLearnedConstraints(ExplodedGraph.Node currentNode) {
-    ExplodedGraph.Node parent = currentNode.parent();
-    if (parent == null) {
-      return Stream.empty();
-    }
-    return currentNode.learnedConstraints()
-      .filter(lc -> lc.symbolicValue().equals(symbolicValue))
-      .filter(this::learnedConstraintForDomains)
-      .map(LearnedConstraint::constraint)
-      .peek(lc -> learnedConstraintFlow(lc, currentNode, parent).forEach(flow::add));
-  }
+  private class ExecutionPath {
+    @Nullable final Symbol trackSymbol;
+    final ExplodedGraph.Edge lastEdge;
+    final PSet<ExplodedGraph.Edge> visited;
+    final List<JavaFileScannerContext.Location> flow;
+    final boolean finished;
 
-  private boolean learnedConstraintForDomains(LearnedConstraint lc) {
-    Constraint constraint = lc.constraint;
-    return constraint == null || domains.stream().anyMatch(d -> d.isAssignableFrom(constraint.getClass()));
-  }
+    ExecutionPath(@Nullable ExplodedGraph.Edge edge, PSet<ExplodedGraph.Edge> visited, @Nullable Symbol trackSymbol, List<JavaFileScannerContext.Location> flow, boolean finished) {
+      this.trackSymbol = trackSymbol;
+      this.lastEdge = edge;
+      this.visited = visited;
+      this.flow = flow;
+      this.finished = finished;
+    }
 
-  private Stream<JavaFileScannerContext.Location> learnedConstraintFlow(@Nullable Constraint learnedConstraint, ExplodedGraph.Node currentNode, ExplodedGraph.Node parent) {
-    if (learnedConstraint == null || !addToFlow.test(learnedConstraint)) {
-      return Stream.empty();
+    ExecutionPath(@Nullable Symbol trackSymbol) {
+      this(null, PCollections.emptySet(), trackSymbol, ImmutableList.of(), false);
     }
-    Tree nodeTree = parent.programPoint.syntaxTree();
-    if (isMethodInvocationNode(parent)) {
-      return methodInvocationFlow(learnedConstraint, currentNode, parent);
+
+    ExecutionPath addEdge(ExplodedGraph.Edge edge) {
+      ImmutableList.Builder<JavaFileScannerContext.Location> flowBuilder = ImmutableList.builder();
+      flowBuilder.addAll(flow);
+      Symbol newTrackSymbol = addFlowFromLearnedAssociations(edge, flowBuilder);
+      Stream<Constraint> constraintStream = addFlowFromLearnedConstraints(edge, flowBuilder);
+      boolean finished = visitedAllParents(edge) || constraintStream.anyMatch(terminateTraversal);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Rename "finished" which hides the field declared at line 126. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AHiddenFieldCheck)